src/backend/access/transam/slru.c

```diff
@@ -617,6 +618,61 @@ SimpleLruWritePage(SlruCtl ctl, int slotno)
 	SlruInternalWritePage(ctl, slotno, NULL);
 }
 
+
+static int
+SimpleLruDownloadSegment(SlruCtl ctl, int pageno, char const* path)
+{
+	int segno;
+	int fd = -1;
+	int n_blocks;
+	char* buffer;
+
+	static SMgrRelationData dummy_smgr_rel = {0};
+
+	/* If page is greather than latest written page, then do not try to download segment from server */
+	if (pageno > ctl->shared->latest_page_number)
+		return -1;
+
+	if (!dummy_smgr_rel.smgr)
+	{
+		RelFileLocator rlocator = {0};
+		dummy_smgr_rel.smgr = smgr(InvalidBackendId, rlocator);
+	}
+	segno = pageno / SLRU_PAGES_PER_SEGMENT;
+
+	buffer = palloc(BLCKSZ * SLRU_PAGES_PER_SEGMENT);
+	n_blocks = smgr_read_slru_segment(&dummy_smgr_rel, path, segno, buffer);
+	if (n_blocks > 0)
```

Comment:
It's a bit wasteful to make the allocation, if there's no smgr hook installed so that smgr_read_slru_segment() does nothing.

Reply:
This code is called only if SLRU segment is missed. It should never happen with Vanilla Postgres.
So  if no smgr hook installed  and SLRU segment is absent, then it is definitely an error. There is no sense to trying to optimise this case.